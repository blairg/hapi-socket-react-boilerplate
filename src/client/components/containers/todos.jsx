/* eslint-disable import/extensions */
/* eslint-disable react/forbid-prop-types */

import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { setTodos } from './../../actions';
import subscribeToTodos from '../../subscribers/todoSubscriber';
import CreateTodo from '../containers/createTodo.jsx';
import Todo from '../containers/todo.jsx';
import calculateTimeDifference from './../../../shared/helpers/timeHelper';

class Todos extends React.Component {
  static getEntries(entries) {
    const entriesNew = [];
    let key = 0;

    entries.map(entry => {
      const updatedEntry = Object.assign({}, entry);
      updatedEntry.displayDate = calculateTimeDifference(
        updatedEntry.timestamp,
      );

      entriesNew.push(<Todo key={key} entry={updatedEntry} />);
      key += 1;

      return updatedEntry;
    });

    return entriesNew;
  }

  componentDidMount() {
    subscribeToTodos(data =>
      this.props.dispatchTodos(data),
    );
  }

  render() {
    return <CreateTodo entries={Todos.getEntries(this.props.entries)} />;
  }
}

Todos.propTypes = {
  entries: PropTypes.array.isRequired,
  dispatchTodos: PropTypes.func.isRequired,
};

const mapDispatchToProps = dispatch => ({
  dispatchTodos: todos => {
    dispatch(setTodos(todos));
  },
});

const mapStateToProps = state => {
  const { todos } = state.setTodos;

  return {
    entries: todos,
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Todos);
