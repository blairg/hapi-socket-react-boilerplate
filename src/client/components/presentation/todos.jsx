/* eslint-disable import/extensions */
/* eslint-disable react/forbid-prop-types */

import React from 'react';
import PropTypes from 'prop-types';

import subscribeToTodos from '../../subscribers/todoSubscriber';
import CreateTodo from '../containers/createTodo.jsx';
import Todo from '../containers/todo.jsx';
import calculateTimeDifference from './../../../shared/helpers/timeHelper';

class Todos extends React.Component {
  static getEntries(entries) {
    const entriesNew = [];
    let key = 0;

    entries.map((entry) => {
      const updatedEntry = Object.assign({}, entry);
      updatedEntry.displayDate = calculateTimeDifference(updatedEntry.timestamp);

      entriesNew.push(<Todo key={key} entry={updatedEntry} />);
      key += 1;

      return updatedEntry;
    });

    return entriesNew;
  }
  constructor(props) {
    super(props);
    this.state = {
      entries: props.entries ? props.entries : [],
    };
  }

  componentDidMount() {
    subscribeToTodos(data => this.setState({
      entries: data,
    }));
  }

  render() {
    const entries = Todos.getEntries(this.state.entries ? this.state.entries : this.props.entries);

    return (
      <CreateTodo entries={entries} />
    );
  }
}

Todos.propTypes = {
  entries: PropTypes.array.isRequired,
};

export default Todos;
