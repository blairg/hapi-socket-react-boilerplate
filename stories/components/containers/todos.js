/* eslint-disable react/jsx-filename-extension */
/* eslint-disable import/no-extraneous-dependencies */

import React from 'react';
import { storiesOf } from '@storybook/react';
import Moment from 'moment';

import Provider from './../../Provider';
import Todos from './../../../src/client/components/containers/todos';

import './../../../public/css/skeleton.css';
import './../../../public/css/site.css';

const timeNow = Moment().unix();
const entries = [
  { title: 'First Title', body: 'First Body', timestamp: timeNow - 10 },
  { title: 'Second Title', body: 'Second Body', timestamp: timeNow - 50 },
  { title: 'Third Title', body: 'Third Body', timestamp: timeNow - 100 },
];
const initialState = {
  setTodos: {
    todos: entries,
  },
};

storiesOf('Todos without data', module)
  .addDecorator(getStory => (
    <Provider story={getStory()} initialState={{}} />
  ))
  .add('without entries', () => <Todos entries={[]} />);

storiesOf('Todos with data', module)
  .addDecorator(getStory => (
    <Provider story={getStory()} initialState={initialState} />
  ))
  .add('with entries', () => <Todos entries={entries} />);
