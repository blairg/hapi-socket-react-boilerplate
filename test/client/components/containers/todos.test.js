/* eslint-disable no-undef */
/* eslint-disable react/jsx-filename-extension */
/* eslint-disable import/extensions */

import React from 'react';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { Todos } from './../../../../src/client/components/containers/todos.jsx';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('./../../../../src/client/subscribers/todoSubscriber');

describe('<Todos />', () => {
  test('should render correctly with data', () => {
    const entries = [{ title: 'my title', body: 'my body' }];
    const wrapper = Enzyme.shallow(<Todos entries={entries} />);

    expect(wrapper).toMatchSnapshot();
  });

  test('should render correctly without data', () => {
    const entries = [];
    const wrapper = Enzyme.shallow(<Todos entries={entries} />);

    expect(wrapper).toMatchSnapshot();
  });
});
