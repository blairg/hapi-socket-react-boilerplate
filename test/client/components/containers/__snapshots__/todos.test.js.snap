// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Todos /> should render correctly with data 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Todos
    entries={
      Array [
        Object {
          "body": "my body",
          "title": "my title",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "entries": Array [
        <Todo
          entry={
            Object {
              "body": "my body",
              "displayDate": "NaN years ago",
              "title": "my title",
            }
          }
        />,
      ],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "entries": Array [
          <Todo
            entry={
              Object {
                "body": "my body",
                "displayDate": "NaN years ago",
                "title": "my title",
              }
            }
          />,
        ],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`<Todos /> should render correctly without data 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Todos
    entries={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "entries": Array [],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "entries": Array [],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
